### PR TITLE
Improve Default Macros

### PR DESCRIPTION
**Pull request by holden-zenith**
Improve Default Macros

Description
-----------

- Cleaned, organized and added documentation to both IPTraceMacroDefaults.h & FreeRTOSIPConfigDefaults.h.
- Removed a few macros that were duplicated in FreeRTOSIPConfigDefaults.h.
- Delete FreeRTOS_errno_TCP.h as all but pdFREERTOS_ERRNO_EAFNOSUPPORT have been included in projdefs.h for quite some time. A PR has been created in [FreeRTOS-Kernel](https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/642/files) to add the last remaining macro. Until then, it resides in FreeRTOSIPConfigDefaults.h.
- Created a new file, FreeRTOSIPDeprecatedDefinitions.h, that contains the checks for usage of deprecated macros. This is similar to FreeRTOS-Kernel's deprecated_definitions.h.
- Added a few missing deprecated macro checks to FreeRTOSIPDeprecatedDefinitions.h.
- Include IPTraceMacroDefaults.h in FreeRTOSIPConfigDefaults.h below the tracing enabling macros and removed all inclusions of IPTraceMacroDefaults.h elsewhere. This ensures the tracing enabling macros are always defined before including the tracing implementation macros.
- Removed some unused deprecated macros (i.e. ipconfigRAND32) from the tests' FreeRTOSIPConfig.h files.

Test Steps
-----------

N/A

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

N/A

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by htibosch — COMMENTED** [review]
Hello @holden-zenith, these changes are very welcome! It's been a long time that `FreeRTOSIPConfigDefaults.h` got cleaned up and better documented, and it was badly needed.
I will read and test your `FreeRTOSIPConfigDefaults.h` with care.
Thanks,

File: source/include/IPTraceMacroDefaults.h
Comment (holden-zenith) [reply source/include/IPTraceMacroDefaults.h:39]: There are some ipconfig macros that enable/disable usage of some of the tracing macros. I just added that requirement to enforce IPTraceMacroDefaults only being imported after all of the config macros have been defined.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (ActoryOu) [reply source/include/FreeRTOSIPConfigDefaults.h:85]: May need to update submodule pointer of FreeRTOS-Kernel in this repository.

**Review by ActoryOu — COMMENTED** [review]
I like the change. Add a minor comment for it.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (htibosch) [reply source/include/FreeRTOSIPConfigDefaults.h:85]: The define is indeed included [here](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/main/include/projdefs.h#L107).

But I would not mind to see a `#warning` in case the macro is unknown: please upgrade your copy of the kernel.

Comment (htibosch) [reply source/include/FreeRTOSIPConfigDefaults.h]: @holden-zenith , I am happy with the new version of FreeRTOSIPConfigDefaults.h It's a lot cleaner and better documented. Thank you.

I read it twice, looking at every detail and made some additions. I will attach my version as a ZIP file: [FreeRTOSIPConfigDefaults.zip](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/files/11128530/FreeRTOSIPConfigDefaults.zip).

Would you be so kind to merge it with your current version? Corrections about the use of English are always welcome.

Note that FreeRTOS originally comes from the UK, and it keeps using the English spelling, as in `behaviour` and `advertise` ( in stead of `behavior` and `advertize` ).

**Comment by holden-zenith** [comment]
I've decided to reduce some of the redundant wording of some documentation and add checks for the values that are expected for each macro. I've done about half of it so far.

**Comment by htibosch** [comment]
@holden-zenith : thank you for merging and editing my suggestions.

> add checks for the values that are expected for each macro

Good idea, thank you

**Comment by moninom1** [comment]
Hi @holden-zenith 
Can you please let us know if you are still working on updating the PR.

**Comment by holden-zenith** [comment]
@moninom1 I had intended to go back to account for all the other code changes but I just haven't had the time to get to it.

**Comment by moninom1** [comment]
@holden-zenith Thank you for the Update.
Can we help in some way to bring this to completion? We have added build separation recently and hence this PR has a lot of conflict. I tried rebasing your change over latest tip but seems like there were a lot more compilation issue. Wanted to check with you first if there are any pending change you are planning to do on this PR, it yes can we move it to draft for now?

Also i see other [PR804](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/804) , [PR747](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/747) and [PR435](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/435) in draft, do you need any help from our side on to review these PRs?

**Comment by holden-zenith** [comment]
@moninom1 I've finally gotten to a point where I might have some time available to bring this PR up to date. [804](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/804) I'm waiting for some hardware to come in to finish. [747](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/747) I'm not sure of how you guys feel about this one but I personally felt it beneficial to merge the files into one and just have a config macro change the method used. Seems beneficial to know which method the user is using and eliminates duplicate code that will fall out of sync.

**Comment by moninom1** [comment]
Hi @holden-zenith 
Thank you for the update. 
We would also like to let you know We will be closing the merging window for dev/IPv6_Integration branch by the end of this month. This is for starting all the verification and testing for the RC3 release and further GA release. 
Can you please let us know if you will be able to update the change before the freezing windows?

**Comment by moninom1** [comment]
@holden-zenith Its great to know. Thank You so much for the update.

**Comment by holden-zenith** [comment]
@moninom1 so here's a question, I had originally defined all macros related to functionality only if that functionality was enabled. For example, ipconfigDNS_CACHE_ENTRIES is only defined if ipconfigUSE_DNS_CACHE is enabled. I did this to help find usage of these macros that aren't guarded by an enabling macro (as they shouldn't need to be checked if the functionality isn't even enabled), but these are what are causing the build errors. Should all macros just be defined regardless or should I go back and define out the areas that shouldn't be/aren't used when certain functionalities are disabled?

**Comment by moninom1** [comment]
HI @holden-zenith 

>> I had originally defined all macros related to functionality only if that functionality was enabled
Ideally this is the right way. 
Regarding ipconfigDNS_CACHE_ENTRIES i can see it is only used in FreeRTOS_DNS_Cache.c which is guarded by both the flags 
```
if ( ( ipconfigUSE_DNS != 0 ) && ( ipconfigUSE_DNS_CACHE == 1 ) )
```
Rest it is only at FreeRTOSIPConfig files. Where is this error coming from for you?

Is ipconfigUSE_DNS_CACHE the only macro giving error? Or there are more such issues?

**Comment by holden-zenith** [comment]
@moninom1 I only gave that as an example, I don't think that was one of the ones giving issues. There were a couple related to DHCP, but I'll check again tomorrow.

Also I tried testing using the added cmake tests but they don't compile due to multiple issues unrelated to my PR so I've gone back to the manual way. Is it just me with this issue?

**Comment by moninom1** [comment]
HI @holden-zenith 

Is there any update on the changes? FYI we have merged the dev/IPv6_integration branch back to main so we can update the change to merge to main.

Thank you

**Comment by moninom1** [comment]
Hi @holden-zenith 
Is there any update on the above PR, we are working on GA release prerequisite and would like to know if the change can be merged for GA. Thank you.

**Comment by holden-zenith** [comment]
@moninom1 Considering this is the type of PR that is best merged after all major changes have been made and all required macros are used/defined, if there are no other major changes planned then I can attempt to resolve the issue with the macros I had mentioned earlier. I worry doing it the 'proper way' could get a little more convoluted than I had planned for this PR as I didn't really want to make any non-trivial code changes outside of the config files but I can check it out and see what will be required now.

**Comment by moninom1** [comment]
Hi @holden-zenith
The Major changes are completed and we are working on bug fixes, We are planning to close the merge window by next Tuesday and as the release is previously committed we are strictly following the dates. We will merge the changes if it is ready by then. Thank you

**Comment by holden-zenith** [comment]
I didn't want to touch too many files in this one PR, so I have reverted unnecessary changes that can be handled at a later time.

File: source/include/FreeRTOS_IP.h
Comment (ActoryOu) [inline source/include/FreeRTOS_IP.h:416]: Is it necessary to remove this condition check?
If so, should we remove the same one in FreeRTOS_TCP_State_Handling.c for the function implementation?

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (ActoryOu) [inline source/include/FreeRTOSIPConfigDefaults.h:2108]:
```suggestion
    #define ipconfigIS_VALID_PROG_ADDRESS( pxAddress )    ( ( pxAddress ) != NULL )
```
I'd suggest to wrap the variable in macro and compare the pointer with NULL instead of 0.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (ActoryOu) [inline source/include/FreeRTOSIPConfigDefaults.h:3030]: Since we merged IPv6 branch back to main, this description is not valid now.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (ActoryOu) [inline source/include/FreeRTOSIPConfigDefaults.h:284]:
```suggestion
    #define ipconfigRA_SEARCH_TIME_OUT_MSEC    ( 10000 )
```

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (ActoryOu) [inline source/include/FreeRTOSIPConfigDefaults.h:309]:
```suggestion
    #define ipconfigRA_IP_TEST_COUNT    ( 3 )
```

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (ActoryOu) [inline source/include/FreeRTOSIPConfigDefaults.h:334]:
```suggestion
    #define ipconfigRA_IP_TEST_TIME_OUT_MSEC    ( 1500 )
```

**Review by ActoryOu — COMMENTED** [review]
Supper great to see the change updating!!
Some minor review comments left.

BTW, do you mind that we check-in any commit to your branch to help fix CI?

**Comment by moninom1** [comment]
Hi @holden-zenith , Thank you for the update, would you please help in moving the changes to **main** instead of **dev/IPv6_Integration** as we have merged the branch back to main, We can also contribute to help fix issues as @ActoryOu  suggested once it is moved to main.

File: source/include/FreeRTOS_IP.h
Comment (holden-zenith) [reply source/include/FreeRTOS_IP.h:416]: Not all of the uses of FreeRTOS_GetTCPStateName were guarded by the condition check, so it's either remove the check from the declaration or fill in the remaining usage checks. This was also causing test failures in FreeRTOS_TCP_State_Handling_utest.c so some of the configFiles for testing would have to be updated.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (holden-zenith) [reply source/include/FreeRTOSIPConfigDefaults.h:284]: Should they all be like this?

Comment (ActoryOu) [reply source/include/FreeRTOSIPConfigDefaults.h:284]: Yes. We'd suggest to use brackets to wrap the macros like guidelines in [FreeRTOS Coding Standard, Testing and Style Guide](https://www.freertos.org/FreeRTOS-Coding-Standard-and-Style-Guide.html).
Below is the example under Style Guide.
`#define A_DEFINITION	( 1 )`

**Comment by holden-zenith** [comment]
@ActoryOu @moninom1 Something that I think is beyond the scope of this PR but has become very noticeable because of it is the widely varying data-types that are used for the same macros that creates the need to cast in many areas. While I think it is unnecessary to make a macro definition for the data-type of each, some effort should probably be made to consolidate this throughout the codebase.

**Comment by ActoryOu** [comment]
Super appreciate your help!

> it is the widely varying data-types that are used for the same macros that creates the need to cast in many areas.

1. I can't understand the first point completely.
Could you help to provide an example of it?



> There's also the issue of not being able to run checks on macros that are casted in the definition via the compiler, so perhaps it would be better to move all checks to vPreCheckConfigs?

2. If the macros are not able to check during compilation, then we can move them to vPreCheckConfigs. Is it possible to list them down?

**Comment by ActoryOu** [comment]
Hi @holden-zenith,
Thanks for explaination.

For the first part, I believe the original code should have casted it, so that we can pass compilation without warnings.
So I suggest that we don't need to handle that in this PR.

For the second part, no worries about the xMPU_SETTINGS issues for now, either. We're skipping those part by `/* LCOV_EXCL_BR_START */` and `/* LCOV_EXCL_BR_STOP */`.

Thanks for the update. I'll try to look into it this week.

**Comment by holden-zenith** [comment]
Thanks, I believe the last issue is just the proofs. I temporarily disabled the FreeRTOS_TCP_Utils_IPV6_ConfigLowTCPMSS_utest because it shouldn't be necessary if a low value is handled by a compile error now, and the FreeRTOS_Sockets_DiffConfig1 tests which use test_FreeRTOS_socket_TCPv6SocketLowMSS.

Also, in regards to the first part in the last comment, I was just mentioning that the down-casting would produce unforeseen issues if the config values were not limited to that data-type size. But I'll just let someone else handle that later on.

**Comment by ActoryOu** [comment]
Hi @holden-zenith,
I've fixed unit-test and proof_ci in your repo.
Here is the summary of my change:
1. Remove length check in FreeRTOS_Sockets.c and FreeRTOS_TCP_Utils_IPv6.c.
    - ipconfigTCP_MSS is guaranteed to be not less than 586 in FreeRTOSIPConfigDefault.h, so it's unnecessary to check that in runtime.
3. Remove redundant unit test cases. (Since ipconfigTCP_MSS is guaranteed to be not less than 586)
4. Set ipconfigTCP_MSS to minimum in some proof_ci test cases.
    - proof_ci is the test to test code with any possible input, it's fine to set ipconfigTCP_MSS greater than network MTU size in this test.

File: source/FreeRTOS_TCP_Utils_IPv6.c
Comment (moninom1) [inline source/FreeRTOS_TCP_Utils_IPv6.c:85]: I think we should have this check in place, as MTU size can be updated according to user. [Example](https://www.freertos.org/FreeRTOS-Plus/FreeRTOS_Plus_TCP/TCP_IP_Configuration_Examples.html)

Comment (ActoryOu) [reply source/FreeRTOS_TCP_Utils_IPv6.c:85]: It's checked in [FreeRTOSIPConfigDefault.h](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/782/files#diff-577fb46f111095d87818a991acfb003ba427900e2bc7c915a6a856990fe7f5f7R1406-R1408). So it's not necessary to check it again here.

**Comment by moninom1** [comment]
Yeah checked it seems fine. Good cleanup of UTs - FreeRTOS_Sockets_DiffConfig1  and FreeRTOS_socket_TCPv6SocketLowMSS test

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:86]: With this are we trying to remove all 0 and 1 check and use ipconfigENABLE & ipconfigDISABLE?

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:112]: Incorrect Indentation

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:182]: I think this is not needed, line 186 check should be enough

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:162]: I think this is not needed, line 186 check should be enough

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (holden-zenith) [reply source/include/FreeRTOSIPConfigDefaults.h:112]: Aside from that one macro missing in older kernel versions, is there actually a particular minimum required kernel version? I didn't want to enforce an arbitrary minimum

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (holden-zenith) [reply source/include/FreeRTOSIPConfigDefaults.h:86]: I just did that to maintain a standardized 'enable' value in the config files since (x == 1) and (x != 0) are used randomly throughout the source and aren't equivalent statements. But at least with the initial config value checks now you can make sure somebody doesn't put something like 2 when you are expecting it to be either 0 or 1, which would cause inconsistencies in the truth value of the config checks in the source.

Comment (moninom1) [reply source/include/FreeRTOSIPConfigDefaults.h:86]: Yeah it would be good if we can use similar kind of check in the complete TCP code, but that we can take up as a separate effort.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:1403]: Can we add it like 40 as this ( ipSIZE_OF_IPv4_HEADER + ipSIZE_OF_TCP_HEADER ) )
As here it is not clear where 40 is coming from

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [reply source/include/FreeRTOSIPConfigDefaults.h:112]: Can we remove it?

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:1138]: I see a lot of such comments are being removed, can we add with the update as well, it will be useful for the end users to understand the macro better. Let me know what do you think about it.

**Comment by moninom1** [comment]
Hi @holden-zenith 
I have left some comments, will you be updating accordingly? If you are fine with the comments let us know, i can create a patch for it. Thank you

**Comment by moninom1** [comment]
Hi @holden-zenith , I have reviewed and updated a commit for some updates in MACRO comments. I will now work on running coverity as pushing if any change needed accordingly.
Meanwhile can you please you please cross check the change and help in rebasing your tip to main. Thank you

**Comment by holden-zenith** [comment]
@moninom1 I am on vacation for the next couple of weeks but I will do whatever is needed afterwards.

**Comment by moninom1** [comment]
Hi @holden-zenith 
I tried rebasing the PR to main, but as a lot of changes have merged from the last rebase, it has become a bit complex to rebase. Also there are a lot of whitespace changes in most of the c files which i do not think are needed now. It would be much easier to create a new patch on the latest tip with just the required changes {Skipping all the whitespace changes}. I can help you with that, but if you want to do that so that the commits are under your name we are fine with it.
Let us know what do you think about it.

**Comment by holden-zenith** [comment]
@moninom1 uncrustify ran for some reason without being intentionally triggered. Just need to revert that. I never really made any changes to the source files.

**Comment by moninom1** [comment]
Hi @holden-zenith 
I tried to rebase this [branch in my repo ](https://github.com/moninom1/FreeRTOS-Plus-TCP/commits/dev-defines).
But as a lot of changes have went in already [mainly [this PR](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/commit/eed294c0ec45433c85597623375a5dbf69058980) ] the change is becoming very big, i would prefer we add the commits over current tip to reduce the amount of changes and keep the PR clean and to the point. Please let me know what do you think.

**Comment by holden-zenith** [comment]
I removed some stuff, it should be easy to rebase now

**Comment by moninom1** [comment]
Yes i think cleaner solution would be to create a new PR. Let me know if you need any help. Thank you

**Comment by moninom1** [comment]
Hi @holden-zenith 

I see the PR is closed. Are you working on the new PR?

**Comment by holden-zenith** [comment]
Sorry yes, I will get to it soon.

File: source/include/FreeRTOSIPConfigDefaults.h
Comment (moninom1) [inline source/include/FreeRTOSIPConfigDefaults.h:246]: Hey @holden-zenith 
Is there any reason we have the commented if at multiple places? MISRA 4.4 forbids to have comments to look like code

Comment (holden-zenith) [reply source/include/FreeRTOSIPConfigDefaults.h:246]: oh yeah you can, they were there to remove configuration definitions that shouldn't be needed if something else is disabled but it turns out that is a little too messy to do at the moment with the current code. Maybe it can be looked at in a later PR.

Comment (moninom1) [reply source/include/FreeRTOSIPConfigDefaults.h:246]: Sure. May be I will remove whatever MISRA4.4 is complaining for now.

**Review by moninom1 — APPROVED** [review]

**Review by ActoryOu — DISMISSED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by htibosch — APPROVED** [review]
Hello @holden-zenith, you have done a great job! I love the formatting, which is now uniform for each config item. Now it also checks if the macro values are valid. It also introduces new names:
`ipconfigENABLE` and `ipconfigDISABLE`, previously known as `1` and `0`.

I have very few remarks, which I uploaded as:

[changes_hein.patch](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/files/13162708/changes_hein.patch)

The use of `ipFALSE_BOOL` is necessary to prevent a MISRA warning which says that `0` is not a boolean value. This expression is accepted:
~~~
#define ipFALSE_BOOL    ( 1 == 2 )
~~~



which can be summarised as:

~~~diff
- * Set to 1 if you want to receive eNetworkDown notification via vApplicationIPNetworkEventHook() callback.
+ * Set to 1 if you want to receive eNetworkDown notification via vApplicationIPNetworkEventHook_Multi() callback.

+ * Function prototype if ipconfigIPv4_BACKWARD_COMPATIBLE is enabled:
+ *
  * eDHCPCallbackAnswer_t xApplicationDHCPHook( eDHCPCallbackPhase_t eDHCPPhase,
  *                                             uint32_t ulIPAddress )
+ *
+ * Function prototype by default:
+ *
+ * eDHCPCallbackAnswer_t xApplicationDHCPHook_Multi( eDHCPCallbackPhase_t eDHCPPhase,
+ *                                                   struct xNetworkEndPoint * pxEndPoint,
+ *                                                   IP_Address_t * pxIPAddress );

-        #define FreeRTOS_debug_printf( MSG )    do {} while( pdFALSE )
+        #define FreeRTOS_debug_printf( MSG )    do {} while( ipFALSE_BOOL )

-        #define FreeRTOS_printf( MSG )    do {} while( pdFALSE )
+        #define FreeRTOS_printf( MSG )    do {} while( ipFALSE_BOOL )

-    #define FreeRTOS_flush_logging()    if( ipconfigHAS_PRINTF || ipconfigHAS_DEBUG_PRINTF ) do {} while( pdFALSE )
+    #define FreeRTOS_flush_logging()    if( ipconfigHAS_PRINTF || ipconfigHAS_DEBUG_PRINTF ) do {} while( ipFALSE_BOOL )

-    #define ipconfigWATCHDOG_TIMER()    do {} while( pdFALSE )
+    #define ipconfigWATCHDOG_TIMER()    do {} while( ipFALSE_BOOL )
~~~

I have compiled and run an IPv4/IPv6 demo on a WIN32 platform. Everything compiled and worked as expected.

I approve the PR and I want thank you a lot for your patience!

**Comment by moninom1** [comment]
Hi @holden-zenith 
I have updated the changes according to Hein's comment, can you please take a look?

**Review by moninom1 — DISMISSED** [review]

**Review by ActoryOu — DISMISSED** [review]

**Comment by moninom1** [comment]
Hi @holden-zenith , I do not understand the reason of dismissing the review request? It was just rebased. Do you see any issue?

**Comment by holden-zenith** [comment]
@moninom1 Seems like that is due to a new change in github [here](https://github.blog/changelog/2023-06-06-security-enhancements-to-required-approvals-on-pull-requests/)

**Review by moninom1 — APPROVED** [review]

**Comment by ActoryOu** [comment]
Hi @holden-zenith,
Do you have any clue to solve this? It seems like system dismisses our approvals automatically.
Or could you help to re-create PR?

Thanks.

**Comment by holden-zenith** [comment]
Seems other people are experiencing the same issue [here](https://github.com/orgs/community/discussions/58535), I tried one of their solutions so let's see if it works.

**Review by moninom1 — APPROVED** [review]

**Review by ActoryOu — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Comment by moninom1** [comment]
Hi All, Please do not merge the change, i am seeing issue in Jenkins run, will check them and update in a while. Thanks

**Comment by moninom1** [comment]
The above change is merged, so merging this as well. Thank you

**Pull request merged**
